### ISSUE TITLE
Bug: [no-redundant-type-constituents] Counterproductive when used with other unions

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#5407]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=C4TwDgpgBAysBOBLAdgcwM5QLxQEQGMB7QgG1ygB88AjCAQ2XVwChRIoAFJAW0WEQBuETDnQIUqSlGQBXbrXgBuZq3DQAwoXkoIAE2yxxaTFS6Je-IekVA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNYmvAIYtaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
type Symbols = "cool" | "beans"
type Primitives = string | number;

type Combined = Symbols | Primitives;
// => "cool" | "beans" is overridden by string in this union type. 4:17 - 4:24
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-redundant-type-constituents": "error"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The rule [I suggested](https://github.com/typescript-eslint/typescript-eslint/issues/3926) is useful when used with simple types, like `string | "arugula"`, but when types come from pre-existing unions it suggests dropping them:

```ts
type Symbols = "cool" | "beans" 
type Primitives = string | number;

type Combined = Primitives; // ✅ 
```

the problem is that now `Combined` is no longer connect to `Symbols`, which may change over time to include types that aren't covered by `Primitives`

### Actual Result

```
"cool" | "beans" is overridden by string in this union type. 4:17 - 4:24
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.31.0` |
| `TypeScript`                           | `4.7.2` |
| `ESLint`                               | `8.15.0` |

**Comment by JoshuaKGoldberg** [comment]
Amusing. Your own creation is now rebelling against you, @fregante!

I think it seems reasonable to consider it a bug that the rule complains in cases where removing one of the type constituents would change the meaning of the intersection or union.

**Comment by fregante** [comment]
> Amusing. Your own creation is now rebelling against you, @fregante!

You have no idea of how many rules I suggested, tried, and disabled 😂 🤭 sorry



> removing one of the type constituents would change the meaning

That is indeed a bug, but if possible it'd be useful to _also_ ignore the first situation I described, unless a deep type is `unknown` or `any` which just cancels out the entire type.

**Comment by JoshuaKGoldberg** [comment]
> _also_ ignore the first situation I described

Ah, sorry, I misinterpreted the intent on that one. I'm not sold that the situation is one where the rule shouldn't complain:

```ts
type Symbols = "cool" | "beans"
type Primitives = string | number;

type Combined = Symbols | Primitives; // ❌ 😳 
// => "cool" | "beans" is overridden by string in this union type. 4:17 - 4:24
```

It seems much more likely to me that a piece of code that _has no effect_ (the `Symbols |`) is a programmer error than any other possibility. It's not generally in the lint rules' purview to guess at what folks are going to do eventually. This strikes me as a very rare case that would be better served by an inline `// eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents` comment.

**Comment by fregante** [comment]
> a piece of code that _has no effect_

True, but that may be temporary. The rule suggest dropping a connection between two nodes only because it (currently) does nothing.

Do you feel the same about this noop operation?

```js
const guests = 10;
const plusOnes = 0;

const total = guests + plusOnes;
// => Adding 0 has no effect. 4:17 - 4:24
```

**Comment by JoshuaKGoldberg** [comment]
Hmm, that's actually not a bad idea for a lint rule... 😂 

Proposal: how about we split this issue in two? Let this one be accepting PRs for a bug fix, and a new, separate one for a feature request to add in an option? You're making good points and it'd be good to get more community feedback on it.
